zio-http/src/main/scala/zio/http/Response.scala

```diff
@@ -1,120 +1,259 @@
+    val status: Status,
+  ) extends Response { self =>
+
+    override final def copy(status: Status, headers: Headers, body: Body): Response =
+      new SocketAppResponse(body, headers, socketApp0, status) with InternalState {
+        override val parent: Response = self
+      }
+
+    override final def freeze: Response = new SocketAppResponse(body, headers, socketApp0, status) with InternalState {
+
+      override val parent: Response = self
+
+      override def frozen: Boolean = true
+
+    }
+
+    override final def updateHeaders(update: Headers => Headers): Response = copy(headers = update(headers))
+
+    override final def withServerTime: Response = new SocketAppResponse(body, headers, socketApp0, status)
+      with InternalState {
+
+      override val parent: Response = self
+
+      override def serverTime: Boolean = true
+    }
+
+  }
+
+  private[zio] class ErrorResponse(val body: Body, val headers: Headers, val httpError0: HttpError, val status: Status)
+      extends Response { self =>
+
+    override def copy(status: Status, headers: Headers, body: Body): Response =
+      new ErrorResponse(body, headers, httpError0, status) with InternalState {
+        override val parent: Response = self
+      }
+
+    override def freeze: Response = new ErrorResponse(body, headers, httpError0, status) with InternalState {
+      override val parent: Response = self
+      override def frozen: Boolean  = true
+    }
+
+    override final def updateHeaders(update: Headers => Headers): Response = copy(headers = update(headers))
+
+    override final def withServerTime: Response = new ErrorResponse(body, headers, httpError0, status)
+      with InternalState {
+
+      override val parent: Response = self
+
+      override def serverTime: Boolean = true
+    }
+  }
+
+  // TODO: This needs to eventually be implementation agnostic.
+  private[zio] class NettyResponse(
+    val body: Body,
+    val channelContext: ChannelHandlerContext,
+    val headers: Headers,
+    val status: Status,
+  ) extends CloseableResponse { self =>
+
+    override final def close(implicit trace: Trace): Task[Unit] = NettyFutureExecutor.executed(channelContext.close())
+
+    override final def copy(status: Status, headers: Headers, body: Body): Response =
+      new NettyResponse(body, channelContext, headers, status) with InternalState {
+        override val parent: Response = self
+      }
+
+    override final def freeze: Response = new NettyResponse(body, channelContext, headers, status) with InternalState {
+      override val parent: Response = self
+      override def frozen: Boolean  = true
+    }
+
+    override final def updateHeaders(update: Headers => Headers): Response = copy(headers = update(headers))
+
+    override final def withServerTime: Response = new NettyResponse(body, channelContext, headers, status)
+      with InternalState {
+      override val parent: Response = self
+
+      override def serverTime: Boolean = true
```

Comment:
Some effort here and in request / body is made to preserve the structure of the original class. But it could be easier for maintenance and not really slower in general (due to happy path) just defaulting to a generic case class that has "all possible fields". Then `copy` on `Response` would just always return that, and it could be overriden selectively if useful in some happy paths. Would keep code much smaller.